proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,521 @@
+
+## Alternatives considered
+
+### Callee-side property wrapper application
+
+Instead of initializing the backing property-wrapper using the argument at the call-site of a function that accepts a wrapped parameter, another approach is to initialize the backing property-wrapper using the parameter in the function body. One benefit of this approach is that annotating a parameter with a property-wrapper attribute would not change the type of the function, and therefore adding or removing a wrapper attribute would be a resilient change.
+
+Under these semantics, using a property-wrapper parameter is effectively the same as using a local property-wrapper that is initialized from a parameter. This implies that:
+
+1. A property-wrapper parameter cannot be used to opt into property-wrapper syntax in the body of a closure that has a parameter with a property-wrapper type.
+2. The type of the argument provided at the call-site cannot affect the overload resolution of `init(wrappedValue:)`.
+3. This feature cannot be extended to allow the call-site to initialize the backing wrapper using a mechanism other than `init(wrappedValue:)`, which is later discussed as a future direction. This further implies that property-wrapper parameters can only be used with property wrappers that support `init(wrappedValue:)`.
+
+One of the main use cases for property-wrapper parameters is opting into property-wrapper syntax in the body of a closure, which makes this approach unviable.
+
+### Property-wrapper attributes as type attributes
+
+One approach for marking closure parameters as property wrappers is to allow property-wrapper custom-attributes to be applied to types, such as:
+
+```swift
+func useReference(
+  _ closure: (@Reference Int) -> Void
+) { ... }
+
+
+useReference { reference in
+  ...
+}
+```
+
+This approach enables inference of the wrapper attribute on the closure parameter from context. However, this breaks the property-wrapper declaration-model, and would force callers to use the property-wrapper syntax. This approach, also, raises questions about anonymous closure-parameters that have an inferred property-wrapper custom-attribute. That is, if an anonymous closure parameter `$0` has the `wrappedValue` type, accessing the backing wrapper and projected value would naturally use `_$0` and `$$0`, which are _far_ from readable. Furthermore, suppose `$0` is bound to the backing wrapper-type; this would mean that naming the parameter would cause the value to change types, which would be very unexpected from a user standpoint. All in all, the property-wrapper syntax is purely an implementation detail for the closure body, which does _not_ belong to the API signature.
+
+
+## Future directions
+
+### Support property-wrapper initialization from a projected value
```

Comment:
The hyphenation of property wrapper is inconsistent throughout the proposal and it's a little distracting. I think we should not hyphenate it at all - it isn't hyphenated anywhere in the original proposal or in TSPL, even when using it as an adjective, e.g. "property wrapper type". For example, in the [original proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0258-property-wrappers.md#proposed-solution):

> This implements the property `foo` in a way described by the _property wrapper type_ for Lazy:

Reply:
I think we should use [this Grammarly article](https://www.grammarly.com/blog/hyphen/) as a guide. Maybe in the following, a hyphen isn't required in the "custom-attributes" part:

```
property-wrapper custom-attributes
                       ^ This might not be necessary
```

However, when using "property wrappers" as an adjective, it _is_ required. For example, saying a "to do list" is kind of confusing, but definitely ungrammatical. Likewise, saying a "property wrapper attribute" is also ungrammatical. It would be valid if "property" and "wrapper" were both used as adjectives and if the former modified the latter. However, that's not the case; it's like they go together. Thus the correct usage of "property wrappers" when used as an adjective is with a hyphen, such as in "property-wrapper type/attribute".

Reply:
Yeah, I understand that it's grammatically correct to hyphenate when "property-wrapper" is used as an adjective. Let's at least remove the hyphenation on nouns, e.g. the phrases "backing property-wrapper" without "type" after, "function-parameter," and "custom-attribute." I'll also fix the hyphenation in the section I just added about memberwise initialization!

Reply:
Yeah, right 😅; I might have overdone it with the nouns, in which (I think) hyphens are not necessary. So, I agree we'd better leave "property wrappers" without any hyphens, when used as a noun. Likewise, we can change the examples you mentioned above. I'll also check for hyphenation on adjectives, so that we catch any errors.